### PR TITLE
XW-217 | verbatim tag replacements: Twitter parser tag

### PR DESCRIPTION
**Pull request by Warkot**

### CODE REVIEW

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	/**
+	 * Sets the href param for <a> and its contents based on the arrtibutes if possible.
+	 *
+	 * @param array $attributes
+	 *
+	 * @return string - rendered HTML markup
+	 */
+	private function renderLink( $attributes ) {
```
Comment (hakubo) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php]: this is hard to follow :(
maybe at least some \n here and there?

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	/**
+	 * Returns a script tag which gets full Twitter script. That second script swaps <a> Twitter tags into widgets.
+	 *
+	 * @return string
+	 */
+	private function getInitializationScript() {
+		return <<<EOT
+<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+"://platform.twitter.com/widgets.js";fjs.parentNode.insertBefore(js,fjs);}}(document,"script","twitter-wjs");</script>
```
Comment (hakubo) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php]: php.js

can this live in a file? or be fetched from other place

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	 * Validates, prefixes and sanitizes the provided attributes.
+	 *
+	 * @param array $attributes - attributes to validate
+	 * @param array $permittedAttributes - key-value pairs of permitted parameters and regexes which these parameters'
+	 *     values have to match.
+	 * @param array $nonPrefixedAttributesNames - attributes which shouldn't get the 'data-' prefix
+	 *
+	 * @return array
+	 */
+	private function prepareAttributes( $attributes, $permittedAttributes, $nonPrefixedAttributesNames ) {
```
Comment (bognix) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php]: how about unit tests for this method?

**Comment by bognix** [comment]
LGTM

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	/**
+	 * Validates, prefixes and sanitizes the provided attributes.
+	 *
+	 * @param array $attributes - attributes to validate
+	 * @param array $permittedAttributes - key-value pairs of permitted parameters and regexes which these parameters'
+	 *     values have to match.
+	 *
+	 * @return array
+	 */
+	private function prepareAttributes( array $attributes, array $permittedAttributes ) {
```
Comment (bognix) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php:88]: consider adding unit tests to this method

File: extensions/wikia/TwitterTag/TwitterTag.i18n.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+$messages = [ ];
+
+$messages['en'] = [
+	'twittertag-desc' => 'Creates the <nowiki><twitter></nowiki> tag.',
```
Comment (vforge) [inline extensions/wikia/TwitterTag/TwitterTag.i18n.php]: please dasherize extension's i18n name

**Comment by vforge** [comment]
LGTM

**Comment by tomasznapieralski** [comment]
LGTM, what about unit tests?
